File: app/src/main/java/com/genymobile/transferclient/MainActivity.java
package com.genymobile.transferclient;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.SurfaceTexture;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.genymobile.transferclient.tools.FileUtils;
import com.genymobile.transferclient.tools.RunProcess;
import com.genymobile.transferclient.tools.ScreenUtil;
import com.genymobile.transferclient.tools.ScreenUtils;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileDescriptor;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.LinkedBlockingQueue;

public class MainActivity extends Activity implements TextureView.SurfaceTextureListener {

    private TextureView textureView;
    private int decoderHeight;
    private int decoderWidth;
    private int dynamicPort;
    private String host;
    private String type;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ScreenUtil.changeFullScreen(this, true);

        Intent intent = getIntent();
        host = intent.getStringExtra("host");
        dynamicPort = intent.getIntExtra("dynamicPort", -1);
        if (dynamicPort == -1) finish();
        type = intent.getStringExtra("type");
        if (type.equals("appRelay")) {
            int realWidth = ScreenUtils.getScreenWidth(this);
            int realHeight = ScreenUtils.getScreenHeight(this);

            decoderWidth = realWidth % 2 != 0 ? realWidth - 1 : realWidth;
            decoderHeight = realHeight % 2 != 0 ? realHeight - 1 : realHeight;

        } else {
            decoderWidth = intent.getIntExtra("width", 0);
            decoderHeight = intent.getIntExtra("height", 0);
        }

        Log.d(TAG, "onCreate: width=" + decoderWidth);
        Log.d(TAG, "onCreate: decoderHeight=" + decoderHeight);

        textureView = new TextureView(this);
        textureView.setSurfaceTextureListener(this);
        FrameLayout layout = new FrameLayout(this);
//        layout.setGravity(Gravity.CENTER);
        layout.addView(textureView);
        FrameLayout.LayoutParams layoutParams = (FrameLayout.LayoutParams) textureView.getLayoutParams();
        layoutParams.width = decoderWidth;
        layoutParams.height = decoderHeight;
        textureView.setLayoutParams(layoutParams);

        setContentView(layout);


    }

    private final LinkedBlockingQueue<MotionEvent> motionEvents = new LinkedBlockingQueue<>();

    private static final String TAG = "MainActivity";


    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onSurfaceTextureAvailable(@NonNull SurfaceTexture surface, int width, int height) {
        Log.d(TAG, "onSurfaceTextureAvailable: ");


        new Thread(() -> {
            try {
                Socket videoSocket = new Socket(host, dynamicPort);
                Log.d(TAG, "socket connection finish " + dynamicPort);
                InputStream inputStream = videoSocket.getInputStream();
                FileDescriptor fileDescriptor = ParcelFileDescriptor.fromSocket(videoSocket).getFileDescriptor();

                VideoDecoder videoDecoder = new VideoDecoder(new Surface(surface),
                        decoderWidth,
                        decoderHeight,
                        fileDescriptor, inputStream);

                Socket controlSocket = new Socket(host, dynamicPort);
                Log.d(TAG, "onSurfaceTextureAvailable: controlSocket");

                DataOutputStream oos = new DataOutputStream(controlSocket.getOutputStream());
                DataInputStream ois = new DataInputStream(controlSocket.getInputStream());
                textureView.setOnTouchListener((v, event) -> {
                    motionEvents.offer(event);
                    return true;
                });
                new Thread(() -> {
                    while (true) {
                        try {
                            MotionEvent event = motionEvents.take();
                            try {
                                oos.writeInt((int) (event.getDownTime()));
                                oos.writeByte((byte)event.getAction());
                                oos.writeByte((byte)event.getPointerCount());

                                for (int i = 0; i < event.getPointerCount(); i++) {
                                    oos.writeByte(event.getPointerId(i));
                                    oos.writeFloat(event.getX(i));
                                    oos.writeFloat(event.getY(i));
                                }
                                oos.flush();
                            } catch (IOException e) {
                                throw new RuntimeException(e);
                            }
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }

                    }
                }).start();

            } catch (Exception e) {
                e.printStackTrace();
            }

        }).start();
    }

    @Override
    public void onSurfaceTextureSizeChanged(@NonNull SurfaceTexture surface, int width, int height) {

    }

    @Override
    public boolean onSurfaceTextureDestroyed(@NonNull SurfaceTexture surface) {
        return false;
    }

    @Override
    public void onSurfaceTextureUpdated(@NonNull SurfaceTexture surface) {

    }

//    public static int packCoordinates(int x, int y) {
//        return (x & 0xFFFF) | ((y & 0xFFFF) << 16);
//    }

}

File: app/src/main/java/com/genymobile/myapplication/MainActivity.kt
package com.genymobile.myapplication

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.os.ParcelFileDescriptor
import android.view.Surface
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.annotation.RequiresApi
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import com.genymobile.myapplication.ui.theme.MyApplicationTheme
import java.io.DataInputStream
import java.io.DataOutputStream
import java.net.ServerSocket
import java.nio.ByteBuffer


class MainActivity : ComponentActivity(), SurfaceHolder.Callback {
    val name = "scrcpy"
    var test_surfaceView: SurfaceView? = null
    var surface_holder: SurfaceHolder? = null
    var surface: Surface? = null
    var d_width: Int? = null
    var d_height: Int? = null
    var thread2_server: Thread? = null

    var videoDecoder: VideoDecoder? = null


    @SuppressLint("ClickableViewAccessibility")
    @RequiresApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var temp = this
        window.setFlags(
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN
        )
        window.decorView.systemUiVisibility =
            View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or View.SYSTEM_UI_FLAG_FULLSCREEN
        setContent {
            MyApplicationTheme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    BoxWithConstraints {
                        val width = maxWidth
                        val height = maxHeight
                        AndroidView(
                            modifier = Modifier
                                .size(width, height),
                            factory = { context ->
                                SurfaceView(context).apply {
                                    layoutParams = ViewGroup.LayoutParams(
                                        (width).value.toInt(),
                                        (height).value.toInt()
                                    )
//                                    setBackgroundColor(0xff000000.toInt())
                                }.also {
                                    test_surfaceView = it
                                    surface_holder = it.holder
                                    // 设置 SurfaceHolder 的回调
                                    // 设置 SurfaceHolder 的回调
                                    surface_holder?.addCallback(temp)

                                }
                            }
                        )
                    }
                }
            }
        }
        thread2_server = Thread {
            try {
//                inputStreamDecoder = InputStreamDecoder(null,1,1)

                val serverSocket = ServerSocket(20001)
                println("server_socket start success!")
//                while (true) {
                try {
                    // videoSocket
                    var videoSocket = serverSocket.accept()
                    var fileDescriptor =
                        ParcelFileDescriptor.fromSocket(videoSocket).fileDescriptor
                    Thread {
                        println("video_socket客户端已连接")
                        var header = ByteBuffer.allocate(/*64 +*/ 4)
                        IO.readFully(fileDescriptor, header)
                        header.flip()

                        val lParam: Int = header.getInt()
                        d_width = lParam shr 16
                        d_height = lParam and 0xFFFF

                        println("$d_width,$d_height")

//                        val name = ByteArray(64)
//                        println("设备名" + header.get(name))

                        videoDecoder = VideoDecoder(
                            fileDescriptor,
                            videoSocket.getInputStream(),
                            surface,
                            d_width!!,
                            d_height!!
                        )
                    }.start()

                    //controlSocket
                    val controlSocket = serverSocket.accept()
                    println("control connect success")
                    val oos = DataOutputStream(controlSocket.outputStream)
                    val ois = DataInputStream(controlSocket.inputStream)
//
//                    var displayId = ois.readLong()
//                    println("display___id $displayId")
                    Thread {
                        println("control_socket客户端已连接")
                        test_surfaceView?.setOnTouchListener { view, event ->
                            println("Pressed at $event")
                            Thread {
//                                oos.writeObject(event)
//                            oos.writeByte(0)


                                oos.writeInt(event.downTime.toInt())
                                oos.writeInt(event.action)
                                oos.writeInt(event.pointerCount)

                                for (i in 0..event.pointerCount - 1) {
                                    oos.writeInt(event.getPointerId(i))
                                    oos.writeInt(event.getX(i).toInt())
                                    oos.writeInt(event.getY(i).toInt())
                                    oos.writeFloat(event.getPressure(i))
                                }
                                oos.writeInt(event.buttonState)
                                oos.writeInt(event.source)
                                oos.flush()
                            }.start()

//                            oos.writeObject(event)
//                            outputStream.writeLong()
                            true
                        }
                    }.start()


                } catch (e: Exception) {
                    e.printStackTrace()
                }
//                }

            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        surface = holder.surface
        thread2_server?.start()
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
    }

}


